Allow legacy unicast clients according to rfc6762#section-6.7

EStrehlau opened the pull request:
Unicast responses already follow https://datatracker.ietf.org/doc/html/rfc6762#section-6.7. 

The only thing missing is supplying the right id for the answer (0 for multicast, id of query for legacy unicast). 
Other than that unicast is only triggered with the unicast bit, instead of also supporting legacy clients. 
This adds detection of legacy clients according to the spec.

EStrehlau committed:
Allow legacy unicast clients according to rfc6762#section-6.7

Unicast responses already follow https://datatracker.ietf.org/doc/html/rfc6762#section-6.7.
The only thing missing is supplying the right id for the answer (0 for multicast, id of query for legacy unicast).
Other than that unicast is only triggered with the unicast bit, instead of also supporting legacy clients.
This adds detection of legacy clients according to the spec.

tenderlove commented:
Excellent, thank you!

EStrehlau committed:
Fix client and fix tests

EStrehlau committed:
Add tests for legacy queries

EStrehlau commented:
I fixed the tests and added some for the legacy behaviour.
Duplication is quite high, I am not sure how to DRY it up since the requirement of separate types for DNS decoding is a bit weird.

tenderlove commented:
> I fixed the tests and added some for the legacy behaviour.
> Duplication is quite high, I am not sure how to DRY it up since the requirement of separate types for DNS decoding is a bit weird.

No worries. As long as we have tests it doesn't matter to me. Ubuntu tests seem to be failing. I'll try to poke at it when I get a chance (if you can get the build green then I'm happy to just merge this though).

EStrehlau commented:
I think those are some flakey tests :S.
I can reproduce failure on linux even with main.
Sadly I do not have time to investigate that much, could you rerun by pull request a few times and see if it fixes things?
I guess same thing goes for the main CI to check if the tests are really flakey or just some issue I have locally.

EStrehlau commented:
NVM, local firewall was broken, but I think I know what is going on:

https://stackoverflow.com/questions/14388706/how-do-so-reuseaddr-and-so-reuseport-differ
> Additionally the kernel performs some "special magic" for SO_REUSEPORT sockets that isn't found in other operating systems: For UDP sockets, it tries to distribute datagrams evenly, for TCP listening sockets, it tries to distribute incoming connect requests (those accepted by calling accept()) evenly across all the sockets that share the same address and port combination. Thus an application can easily open the same port in multiple child processes and then use SO_REUSEPORT to get a very inexpensive load balancing.

I.e. since we now have client and server on 5353 (and both count as the same interface) the server response or the client requests might only right back where it came from.

Disabling SO_REUSEPORT fixes all tests.

I think this might also break if there is a seconds MDNS like avahi running since then packets are again distributed between the two.
This seems like a compilicated topic, avahi seems to set SO_REUSEPORT.
mDNSResponsder seems to prefer only SO_REUSEADDR: https://github.com/apple-oss-distributions/mDNSResponder/blob/d5029b5dff8aa59d1fc07ed796e994106ef58dee/mDNSPosix/mDNSPosix.c#L1126

I did not merge the fix yet so you don't accidentally merge to master when you see all tests are green ;).
I think this warrants an executive decision.

EStrehlau commented:
I did some more digging and not having SO_REUSEPORT on Linux has the same behaviour as SO_REUSEPORT and SO_REUSEADDR set on macOS.

http://hacked10bits.blogspot.com/2014/12/udp-binding-and-port-reuse-in-linux.html?m=1

I think there is no way to have consistent behaviour on multiple platforms by always setting both flags:
https://fabiensanglard.net/mdns/

So I'll just do not set the flag on Linux, this should also work on BSD and Windows I think, only linux seems the odd one out.

EStrehlau committed:
Fix unwanted load balancing between 5353 socket on linux.

http://hacked10bits.blogspot.com/2014/12/udp-binding-and-port-reuse-in-linux.html?m=1
https://stackoverflow.com/questions/14388706/how-do-so-reuseaddr-and-so-reuseport-differ

tenderlove commented:
Blech. Makes sense. Thank you so much for digging in to this.  I'll merge and ship a new version.

tenderlove commented:
@EStrehlau this PR seems to have broken the client on macOS (and Linux) and I'm not sure how to fix it.  Since merging the PR I can't discover anything on my network.  Specifically since de74a8c209df201c46dcd8805239611801a3d826.

I'm going to revert the merge and reopen this PR.

EStrehlau commented:
@tenderlove we depend on this for production.

We are using the library on both macOS and Linux, so I am confused what is triggering this behaviour.
We have not seen it, my changes have especially addressed usage in parallel to avahi/bonjour.
It probably still is a weird interaction with those pieces of software running tho.

Would you we willing to do a tcpdump that includes the non working discovery?
If there are application level issues can you share them.

EStrehlau commented:
@tenderlove NVM, I think those changes are related to the client we do not use.

We do not use the client so it makes sense that something has slipped through.
Could we maybe leave the server changes in and just revert the client changes?

Per spec mDNS clients are supposed to send on the mDNS port. This is used to differentiate between a "legacy" client and a mDNS client on the server side. So I think my changes are actually the "correct" behaviour.

I can probably test to see what the issue is with the zeroconf client in combination with bonjor/avahi and then make some fixes.

Does this sound good to you:

- Revert the revert for server changes.
- Leave the revert for the client for now.
- I'll try to reproduce and try to do a spec conforming fix at some point in the future.

tenderlove commented:
> * Revert the revert for server changes.
> * Leave the revert for the client for now.
> * I'll try to reproduce and try to do a spec conforming fix at some point in the future.

Yes, I'm happy with this.  AFAICT, you were just changing the client in order to test the server (based on bisecting to the particular commit).  I can tackle the first two bullet points, and I'll open an issue for the third.

tenderlove commented:
Hrm, I guess the client tests depend on the implementation of the server.  @EStrehlau I agree with your suggestion, but reverting the revert for the server might be a little challenging if we want to keep the build green (which I do).  I think I can do this, but it might take a little time.

tenderlove commented:
Doing a little debugging, it looks like the server is considering the client to be "legacy" because of [this line](https://github.com/tenderlove/zeroconf/blob/v1.0.1/lib/zeroconf/service.rb#L125).  If I set legacy to `false` the client tests will pass.  Is this the right way to determine if a client should be considered legacy?

If the client code is technically "legacy", I'd expect the service to just handle it (IOW, the client tests should pass without modification to the client code).

Alternatively, we could allow the client port to be configurable, then just run the client on 5353 in test (but use port 0 by default).  Anyway, please let me know what you think.

EStrehlau commented:
This is the corresponding section in the RFC:
https://datatracker.ietf.org/doc/html/rfc6762#section-6.7

So I think a spec conforming mDNS client needs to send from port 5353 instead of a random "ephemeral" port like in the port 0 case.

I really do not understand why this is causing issues. I guess it is probably because other mDNS services are bound to port 5353? My other changes should've accounted for that..., not sure at this point with time passing.

It is important to differentiate between legacy/non legacy clients since the behaviour of the server is supposed to be different.
It looks like the "unicast-response" bit is not enough to figure this out since the RFC has not inverted it.... I.e. a legacy client that does not know about mDNS always says "I want a multicast response". IMHO it should be the other way around where mDNS clients that know about that bit explicitly have to opt into multicast...

I'll do some digging into other mDNS client implementations and see if the spec is just not followed causing interop issues.
Otherwise there must be some magic combination of SO_ flags for client and server to make this all work...

Who would've guessed mDNS is such a mess :D.
I informed my colleagues to pin to the old version, so don't feel pressure to fix this.
I think the legacy client is a more niche usecase over a working client.
Alltough the client in the current (your old) state might be misidentified as a legacy client.

Could you let me know with what server implementations the client has issues with?
Obviosly not yours since the tests are passing.

EStrehlau commented:
Ok I did some digging:
Looks like the mDNS clients are expected to send on sourceport 5353 in practice.
Both Avahi and systemd-resolve treat non 5353 traffic as legacy:
https://github.com/avahi/avahi/blob/1dade81cbf4d3e3f49784e44e543cd77046c79df/avahi-core/server.c#L980
https://github.com/systemd/systemd/blob/c195fdba12a0f8e2a831c27fe73f8afe9125a06a/src/resolve/resolved-mdns.c#L192

I think we are probably expecting a unicast response on the client/setting the unicast bit.
Which we cannot receive since another mDNS Server is listening on 5353 and linux either hand all unicast responses to that server or load balances it between us and the server (lost packets either way)

If I am correct with this we just need to disable the unicast bit in the client packets we send.

tenderlove commented:
> If I am correct with this we just need to disable the unicast bit in the client packets we send.

Am I understanding correctly that the client _should_ be on port 5353, but we need to _not_ set the unicast bits in the client packets?

Also, would you mind sending a PR to implement what you say (as I'm not 100% sure I'm following)?  I can manually test its still working for my use-case and then we can put this patch back in the `main` branch.
